nimbess-cni.go

```diff
@@ -0,0 +1,314 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Main package for the Nimbess CNI plugin
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/version"
+	"google.golang.org/grpc"
+
+	cnitypes "github.com/containernetworking/cni/pkg/types/current"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+	cninimbess "github.com/nimbess/nimbess-agent/pkg/proto/cni"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultLogFile = "/tmp/nimbess-cni.log"
+
+// NimbessConfig is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type nimbessConfig struct {
+	// common CNI config
+	types.NetConf
+
+	// PrevResult contains previous plugin's result, used only when called in the context of a chained plugin.
+	PrevResult *map[string]interface{} `json:"prevResult"`
+
+	// GrpcServer is a plugin-specific config, contains location of the gRPC server
+	// where the CNI requests are being forwarded to (server:port tuple, e.g. "localhost:9111")
+	// or unix-domain socket path (e.g. "/run/cni.sock").
+	GrpcServer string `json:"grpcServer"`
+
+	// LogFile is a plugin-specific config, specifies location of the CNI plugin log file.
+	// If empty, plugin logs into defaultLogFile.
+	LogFile string `json:"logFile"`
+
+	// EtcdEndpoints is a plugin-specific config, may contain comma-separated list of ETCD endpoints
+	// required for specific for the CNI / IPAM plugin.
+	EtcdEndpoints string `json:"etcdEndpoints"`
+}
+
+// grpcConnect sets up a connection to the gRPC server specified in grpcServer argument
+// as a server:port tuple (e.g. "localhost:9111") or unix-domain socket path (e.g. "/run/cni.sock").
+func grpcConnect(grpcServer string) (conn *grpc.ClientConn, cni cninimbess.RemoteCNIClient, err error) {
+
+	if grpcServer != "" && grpcServer[0] == '/' {
+		// unix-domain socket connection
+		conn, err = grpc.Dial(
+			grpcServer,
+			grpc.WithInsecure(),
+			grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+				return net.DialTimeout("unix", addr, timeout)
+			}),
+		)
+	} else {
+		// TCP socket connection
+		conn, err = grpc.Dial(grpcServer, grpc.WithInsecure())
+	}
+
+	if err != nil {
+		return
+	}
+	cni = cninimbess.NewRemoteCNIClient(conn)
+	return
+}
+
+// parseCNIConfig parses CNI config from JSON (in bytes) to nimbessConfig struct.
+func parseCNIConfig(bytes []byte) (*nimbessConfig, error) {
```

Comment:
Guess we should standardize on our config pattern... Tim and I went with yaml configs, though we can use json also since we used a viper library. No need to change this here as I think we will create a config library to use across the projects.

Reply:
I'm just following CNI spec:

The network configuration is in JSON format and can easily be stored in a file. The network configuration includes mandatory fields such as "name" and "type" as well as plugin (type) specific ones. The network configuration allows for fields to change values between invocations. For this purpose there is an optional field "args" which must contain the varying information.